ping/ping_common.c

```diff
@@ -455,20 +455,25 @@ void sock_setbufs(struct ping_rts *rts, socket_st *sock, int alloc)
 	}
 }
 
-void sock_setmark(unsigned int mark, int fd)
+void sock_setmark(struct ping_rts *rts, int fd)
 {
 #ifdef SO_MARK
 	int ret;
 	int errno_save;
 
+	if (!rts->opt_mark)
+		return;
+
 	enable_capability_admin();
-	ret = setsockopt(fd, SOL_SOCKET, SO_MARK, &mark, sizeof(mark));
+	ret = setsockopt(fd, SOL_SOCKET, SO_MARK, &(rts->mark), sizeof(rts->mark));
 	errno_save = errno;
 	disable_capability_admin();
 
-	/* Do not exit, old kernels do not support mark. */
-	if (ret == -1)
-		error(0, errno_save, _("WARNING: failed to set mark: %u"), mark);
+	if (ret == -1) {
+		error(0, errno_save, _("WARNING: failed to set mark: %u"), rts->mark);
+		error(0, 0, _("=> missing cap_net_admin+p capability?"));
```

Comment:
You could limit this hint to cases when errno actually is `EPERM` (or `EACCES`?)

Reply:
Makes sense actually, thanks! I also add `(or run ping under root)`, although it might not be a good idea due security to run ping under root (but it will be needed if user needs it and distro does not add it or he does not modify the capability himself).

Reply:
Current kernel set `EPERM`, therefore I'll use just that. 
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/core/sock.c?h=v6.7#n1415

Also, since v5.17 the check added also `CAP_NET_RAW` (so now it's  `CAP_NET_RAW`  or `CAP_NET_ADMIN`: [079925cce1d0 ("net: allow SO_MARK with CAP_NET_RAW")
](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=079925cce1d06fe6614b1559b0020497b5b83d6a). Therefore I sent a patch to linux man pages https://lore.kernel.org/linux-man/20240116110418.2577798-1-pvorel@suse.cz/ and note it in ping man page as well.